File: src/main/application/score.py
from __future__ import annotations
import logging
from datetime import datetime, timedelta
from collections import Counter

from application.ports import SignalRepository, OpportunityRepository, Notifier
from domain.models import Opportunity, Signal, ActiveSegment
from domain.rules import (
    compute_opportunity_score, income_tier_score, urgency_score, volume_score,
    KILL_SCORE_THRESHOLD, SCALE_SCORE_THRESHOLD, ALERT_SCORE_THRESHOLD,
)

log = logging.getLogger(__name__)


class ScoreUseCase:
    def __init__(self, signal_repo: SignalRepository,
                 opp_repo: OpportunityRepository, notifier: Notifier):
        self._signals = signal_repo
        self._opps = opp_repo
        self._notifier = notifier

    def run(self, segments: list[ActiveSegment], dry_run: bool = False) -> list[dict]:
        results = []

        for seg in segments:
            signals = self._signals.get(segment=seg.key, limit=500)
            opp = self._score_segment(seg, signals)
            opp = self._apply_rules(opp)

            if not dry_run:
                self._opps.upsert(opp)
                if opp.score >= ALERT_SCORE_THRESHOLD and opp.status == "watching":
                    if self._should_alert(opp):
                        msg = self._format_alert(opp, seg)
                        if self._notifier.send(msg):
                            opp.telegram_alerted_at = datetime.utcnow()
                            self._opps.upsert(opp)

            results.append({
                "segment": seg.key,
                "score": opp.score,
                "status": opp.status,
                "signal_count": opp.signal_count,
                "breakdown": opp.score_breakdown,
                "pain_summary": opp.pain_summary,
            })

        results.sort(key=lambda r: r["score"], reverse=True)
        return results

    def _score_segment(self, seg: ActiveSegment, signals: list[Signal]) -> Opportunity:
        dolor, pain_summary = self._dolor_score(signals)
        breakdown = {
            "dolor":          dolor,
            "capacidad_pago": float(income_tier_score(seg.income_tier)),
            "volumen":        volume_score(seg.discovery_score),
            "competencia":    5.0,
            "urgencia":       float(urgency_score(seg.has_deadline)),
        }
        existing = self._opps.get_by_segment(seg.key)
        opp_id = existing.id if existing else Opportunity(segment=seg.key).id
        return Opportunity(
            id=opp_id,
            segment=seg.key,
            pain_summary=pain_summary or (existing.pain_summary if existing else ""),
            score=compute_opportunity_score(breakdown),
            score_breakdown=breakdown,
            signal_ids=[s.id for s in signals[-50:]],
            signal_count=self._signals.count(segment=seg.key),
            first_seen=existing.first_seen if existing else datetime.utcnow(),
            last_updated=datetime.utcnow(),
            status=existing.status if existing else "watching",
            landing_url=existing.landing_url if existing else None,
            emails_captured=existing.emails_captured if existing else 0,
            telegram_alerted_at=existing.telegram_alerted_at if existing else None,
        )

    def _dolor_score(self, signals: list[Signal]) -> tuple[float, str]:
        if not signals:
            return 0.0, ""
        cutoff = datetime.utcnow() - timedelta(days=30)
        recent = [s for s in signals if s.collected_at > cutoff]
        if not recent:
            return 0.0, ""
        week_ago = datetime.utcnow() - timedelta(days=7)
        freq_score = min(len(recent) / 20, 1.0) * 10
        weighted, total_w, all_kw = 0.0, 0.0, []
        for s in recent:
            w = 2.0 if s.collected_at > week_ago else 1.0
            weighted += s.signal_strength * w
            total_w += w
            all_kw.extend(s.pain_keywords_found)
        intensity = (weighted / total_w) * 10 if total_w else 0
        dolor = round((freq_score * 0.5 + intensity * 0.5), 2)
        top_kw = [kw for kw, _ in Counter(all_kw).most_common(3)]
        summary = f"Dolor en: {', '.join(top_kw)}" if top_kw else f"{len(recent)} señales recientes"
        return min(dolor, 10.0), summary

    def _apply_rules(self, opp: Opportunity) -> Opportunity:
        if opp.status in ("killed", "scaling"):
            return opp
        age = (datetime.utcnow() - opp.first_seen).days
        if opp.signal_count == 0 and age >= opp.kill_threshold_days and opp.score < KILL_SCORE_THRESHOLD:
            opp.status = "killed"
        elif opp.score >= SCALE_SCORE_THRESHOLD and opp.emails_captured >= opp.scale_threshold_emails:
            opp.status = "scaling"
        return opp

    def _should_alert(self, opp: Opportunity) -> bool:
        if not opp.telegram_alerted_at:
            return True
        return (datetime.utcnow() - opp.telegram_alerted_at).total_seconds() > 86400

    def _format_alert(self, opp: Opportunity, seg: ActiveSegment) -> str:
        bd = opp.score_breakdown
        lines = [
            f"🎯 *Oportunidad detectada*",
            f"*Segmento:* {seg.label}",
            f"*Score:* {opp.score}/10",
            f"*Dolor:* {bd.get('dolor', 0):.1f} | *Pago:* {bd.get('capacidad_pago', 0):.0f} | *Urgencia:* {bd.get('urgencia', 0):.0f}",
            f"*Señales:* {opp.signal_count}",
            f"*Resumen:* {opp.pain_summary}",
        ]
        if seg.has_deadline:
            lines.append("⚠️ Deadline activo")
        return "\n".join(lines)

File: src/main/infrastructure/worker/score.js
import { sendTelegram } from "./notify.js";

export const SCORE_WEIGHTS = {
  dolor: 0.30,
  capacidad_pago: 0.25,
  volumen: 0.20,
  competencia: 0.15,
  urgencia: 0.10,
};

export const KILL_SCORE_THRESHOLD = 5.0;
export const SCALE_SCORE_THRESHOLD = 8.0;
export const ALERT_SCORE_THRESHOLD = 7.0;

const SALARY_TIERS = { high: 10, medium_high: 7, medium: 5, low: 2 };
const DEFAULT_COMPETENCIA_SCORE = 5.0;

export function computeOpportunityScore(breakdown) {
  const raw = Object.entries(SCORE_WEIGHTS).reduce(
    (sum, [k, w]) => sum + (breakdown[k] ?? 0) * w, 0
  );
  return Math.round(raw * 100) / 100;
}

export function incomeTierScore(tier) {
  return SALARY_TIERS[tier] ?? 2;
}

export function urgencyScore(hasDeadline) {
  return hasDeadline ? 10 : 0;
}

export function volumeScore(discoveryScore) {
  return Math.round(Math.min(discoveryScore / 20.0, 1.0) * 10 * 100) / 100;
}

export function dolorScore(signals) {
  const now = Date.now();
  const cutoff = now - 30 * 86400000;
  const weekAgo = now - 7 * 86400000;
  const recent = signals.filter(s => new Date(s.collected_at).getTime() > cutoff);
  if (!recent.length) return [0, ""];

  const freqScore = Math.min(recent.length / 20, 1.0) * 10;
  let weighted = 0, totalW = 0;
  const allKw = [];

  for (const s of recent) {
    const w = new Date(s.collected_at).getTime() > weekAgo ? 2.0 : 1.0;
    weighted += (s.signal_strength ?? 0) * w;
    totalW += w;
    let kws = [];
    try { kws = JSON.parse(s.pain_keywords || "[]"); } catch { kws = []; }
    allKw.push(...kws);
  }

  const intensity = totalW ? (weighted / totalW) * 10 : 0;
  const dolor = Math.min(Math.round((freqScore * 0.5 + intensity * 0.5) * 100) / 100, 10.0);

  const kwCount = {};
  for (const kw of allKw) kwCount[kw] = (kwCount[kw] ?? 0) + 1;
  const topKw = Object.entries(kwCount).sort((a, b) => b[1] - a[1]).slice(0, 3).map(([k]) => k);
  const summary = topKw.length ? `Dolor en: ${topKw.join(", ")}` : `${recent.length} señales recientes`;

  return [dolor, summary];
}

export function applyRules(opp) {
  if (opp.status === "killed" || opp.status === "scaling") return opp;
  const ageDays = (Date.now() - new Date(opp.first_seen).getTime()) / 86400000;
  if ((opp.signal_count ?? 0) === 0 && ageDays >= (opp.kill_threshold_days ?? 7) && opp.score < KILL_SCORE_THRESHOLD) {
    return { ...opp, status: "killed" };
  }
  if (opp.score >= SCALE_SCORE_THRESHOLD && (opp.emails_captured ?? 0) >= (opp.scale_threshold_emails ?? 30)) {
    return { ...opp, status: "scaling" };
  }
  return opp;
}

export function shouldAlert(opp) {
  if (!opp.telegram_alerted_at) return true;
  return (Date.now() - new Date(opp.telegram_alerted_at).getTime()) > 86400000;
}

export function formatAlert(opp, seg) {
  const bd = opp.score_breakdown ?? {};
  const lines = [
    `🎯 *Oportunidad detectada*`,
    `*Segmento:* ${seg.label}`,
    `*Score:* ${opp.score}/10`,
    `*Dolor:* ${(bd.dolor ?? 0).toFixed(1)} | *Pago:* ${(bd.capacidad_pago ?? 0).toFixed(0)} | *Urgencia:* ${(bd.urgencia ?? 0).toFixed(0)}`,
    `*Señales:* ${opp.signal_count}`,
    `*Resumen:* ${opp.pain_summary}`,
  ];
  if (seg.has_deadline) lines.push("⚠️ Deadline activo");
  return lines.join("\n");
}

export async function runScore(env, topN = 10, minScore = 1.0, dryRun = false) {
  const activeSegments = await _buildActiveSegments(env, topN, minScore);
  const results = [];

  for (const seg of activeSegments.values()) {
    const { results: signals } = await env.DB.prepare(
      "SELECT * FROM signals WHERE segment = ? ORDER BY collected_at DESC LIMIT 500"
    ).bind(seg.key).all();
    const countRow = await env.DB.prepare(
      "SELECT COUNT(*) as n FROM signals WHERE segment = ?"
    ).bind(seg.key).first();

    const [dolor, painSummary] = dolorScore(signals ?? []);
    const breakdown = {
      dolor,
      capacidad_pago: incomeTierScore(seg.income_tier),
      volumen: volumeScore(seg.discovery_score),
      competencia: DEFAULT_COMPETENCIA_SCORE,
      urgencia: urgencyScore(seg.has_deadline),
    };
    const score = computeOpportunityScore(breakdown);

    const existing = await env.DB.prepare(
      "SELECT * FROM opportunities WHERE segment = ? LIMIT 1"
    ).bind(seg.key).first();

    const now = new Date().toISOString();
    const opp = {
      id: existing?.id ?? crypto.randomUUID().slice(0, 8),
      segment: seg.key,
      pain_summary: painSummary || existing?.pain_summary || "",
      score,
      score_breakdown: breakdown,
      signal_ids: (signals ?? []).slice(-50).map(s => s.id),
      signal_count: countRow?.n ?? 0,
      first_seen: existing?.first_seen ?? now,
      last_updated: now,
      status: existing?.status ?? "watching",
      landing_url: existing?.landing_url ?? null,
      emails_captured: existing?.emails_captured ?? 0,
      kill_threshold_days: 7,
      scale_threshold_emails: 30,
      telegram_alerted_at: existing?.telegram_alerted_at ?? null,
    };

    const finalOpp = applyRules(opp);

    if (!dryRun) {
      await env.DB.prepare(`
        INSERT INTO opportunities
          (id, segment, pain_summary, score, score_breakdown, signal_ids,
           signal_count, first_seen, last_updated, status, landing_url,
           emails_captured, validation_deadline, telegram_alerted_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ON CONFLICT(id) DO UPDATE SET
          score=excluded.score, score_breakdown=excluded.score_breakdown,
          signal_ids=excluded.signal_ids, signal_count=excluded.signal_count,
          last_updated=excluded.last_updated, status=excluded.status,
          emails_captured=excluded.emails_captured, landing_url=excluded.landing_url,
          validation_deadline=excluded.validation_deadline,
          telegram_alerted_at=excluded.telegram_alerted_at
      `).bind(
        finalOpp.id, finalOpp.segment, finalOpp.pain_summary ?? null, finalOpp.score,
        JSON.stringify(finalOpp.score_breakdown), JSON.stringify(finalOpp.signal_ids),
        finalOpp.signal_count, finalOpp.first_seen, finalOpp.last_updated,
        finalOpp.status, finalOpp.landing_url ?? null, finalOpp.emails_captured ?? 0,
        null, finalOpp.telegram_alerted_at ?? null,
      ).run();

      if (finalOpp.score >= ALERT_SCORE_THRESHOLD && finalOpp.status === "watching" && shouldAlert(finalOpp)) {
        const msg = formatAlert(finalOpp, seg);
        const sent = await sendTelegram(env, msg);
        if (sent) {
          await env.DB.prepare(
            "UPDATE opportunities SET telegram_alerted_at = ? WHERE id = ?"
          ).bind(now, finalOpp.id).run();
        }
      }
    }

    results.push({
      segment: seg.key, score: finalOpp.score, status: finalOpp.status,
      signal_count: finalOpp.signal_count, breakdown, pain_summary: finalOpp.pain_summary,
    });
  }

  return results.sort((a, b) => b.score - a.score);
}

async function _buildActiveSegments(env, topN, minScore) {
  const segments = new Map();

  const latestRun = await env.DB.prepare(
    "SELECT run_id FROM discovery_candidates ORDER BY id DESC LIMIT 1"
  ).first();

  if (latestRun) {
    const { results: candidates } = await env.DB.prepare(
      `SELECT * FROM discovery_candidates
       WHERE run_id = ? AND discovery_score >= ?
       ORDER BY discovery_score DESC LIMIT ?`
    ).bind(latestRun.run_id, minScore, topN).all();

    for (const c of candidates ?? []) {
      const key = c.profile.toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_|_$/g, "").slice(0, 48);
      segments.set(key, {
        key, label: c.profile,
        keywords: JSON.parse(c.keywords || "[]"),
        income_tier: c.income_est || "medium",
        has_deadline: c.has_deadline === 1,
        discovery_score: c.discovery_score ?? 0,
      });
    }
  }

  const { results: leadSegs } = await env.DB.prepare(
    "SELECT DISTINCT segment FROM leads"
  ).all();
  for (const r of leadSegs ?? []) {
    if (!segments.has(r.segment)) {
      segments.set(r.segment, {
        key: r.segment, label: r.segment, keywords: [],
        income_tier: "medium", has_deadline: false, discovery_score: 0,
      });
    }
  }

  return segments;
}
